### PR TITLE
Add Eval documentation

### PR DESCRIPTION
**Pull request by LukaJCB**
Adds some eval documentation with very basic examples, not sure if there are better examples around that can better showcase the usefulness of `Eval`. Either way, looking forward to feedback :)

### CODE REVIEW

**Comment by johnynek** [comment]
This does not seem to mention that some uses of nested `.value` calls can be unsafe and kill stack safety.

Ideally, you never call `.value` inside something that could be itself inside an Eval. I don't fully grok, myself rules for when it might be safe to call `.value` inside another Eval, but @non seems to have some rules of thumb. I think without this discussion we could be setting someone up for confusion stack overflows.

**Comment by johnynek** [comment]
note for instance #1703 which I think may be currently broken still.

**Comment by johnynek** [comment]
Can we add an example of using `Eval.defer` to do stack safe recursion? That's one of the main motivators of Eval (along with giving a type to an optionally lazy value).

another note: it would be really nice to have some, even non-rigorous, benchmarks on Eval. In fact, Eval is MUCH slower than using standard function calls. So if you are doing it to get stack safe recursion, if you know the recursion depth will be shallow, you may be much better off with normal on-stack recursion.

cc @non

File: docs/src/main/tut/datatypes/eval.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+We can run the computation using the given evaluation strategy anytime by using the `value` method.
+
+```tut:book
+eager.value
+
+```
+
+#### Eval.later
+
+If we want lazy evaluation, we can use `Eval.later`:
```
Comment (kailuowang) [inline docs/src/main/tut/datatypes/eval.md]: Might be worthy mentioning here that there are two advantages of `Later` over `lazy val` 
1. `Later` allows the thunk being GC'd after evaluation and
2. `Later`'s synchronization lock locks itself while lazy vals lock their enclosing class.

**Comment by kailuowang** [comment]
I agree with @ceedubs that having this is much better than having none, @LukaJCB what do you think? Do you think those points are super quick to address or do you want to iterate fast? 
BTW, there is an awesome talk by @non on this topic. https://www.youtube.com/watch?v=XDDy-QD0aVU maybe worth mentioning

**Comment by LukaJCB** [comment]
I did address @kailuowangs points and I also added a small note on `Eval.defer` :) 
I was going to wait for @non to chime in here, about stack-safe recursion, since I'm not super knowledgable on that, but I'm fine with merging now, and making a new PR with `Eval.defer`.

**Comment by ceedubs** [comment]
Merging with 2 approvals. I do think that @kailuowang's suggestion to link to Erik's talk on `Eval` would be a nice addition in a future PR.

**Pull request merged**